File: packages/core_dart/test/publisher_profile_test.dart
import 'package:kromia_core/kromia_core.dart';
import 'package:test/test.dart';

/// KRO-285 / KRO-295 — espejo Dart del perfil público. Se prueban las reglas
/// que, espejadas mal, se notan tarde: los huecos que son valores de verdad y el
/// conjunto cerrado de colores.
void main() {
  group('a quién apunta un álbum', () {
    test('un publisher trae slug con el que pedir su perfil', () {
      final r = AlbumCreatorRef.fromJson(const {
        'kind': 'publisher',
        'ref': 'forjabrasa',
        'displayName': 'Forjabrasa Ediciones',
      });
      expect(r.hayAlguien, isTrue);
      expect(r.ref, 'forjabrasa');
    });

    test('`unknown` NO trae ref, y eso es el contrato', () {
      // Mandar a la app a una petición que va a fallar es peor que decirle desde
      // el principio que no hay nadie.
      final r = AlbumCreatorRef.fromJson(const {'kind': 'unknown'});
      expect(r.ref, isNull);
      expect(r.hayAlguien, isFalse);
    });

    test('un ref en blanco cuenta como que no hay nadie', () {
      expect(
        AlbumCreatorRef.fromJson(const {'kind': 'publisher', 'ref': '   '}).hayAlguien,
        isFalse,
      );
    });

    test('un kind que no conocemos degrada a desconocido, no a pantalla vacía', () {
      // Seguir el ref de un tipo que la app no sabe abrir acaba en blanco; la
      // pantalla del desconocido es la degradación segura.
      final r = AlbumCreatorRef.fromJson(const {'kind': 'sindicato', 'ref': 'x'});
      expect(r.hayAlguien, isFalse);
    });
  });

  group('banda de color', () {
    test('el conjunto es CERRADO y neutral es el default', () {
      expect(profileBandColors, contains('neutral'));
      expect(defaultBandColor, 'neutral');
      expect(isBandColor('fucsia'), isFalse);
      expect(isBandColor('gold'), isTrue);
    });

    test('cada banda trae su color de TEXTO, no solo el fondo', () {
      // El `fg` viene del contrato para que el nombre se lea encima; dejarlo al
      // criterio de cada host es cómo aparece texto blanco sobre oro.
      for (final c in profileBandColors) {
        expect(bandColorValues[c]!.bg, startsWith('#'));
        expect(bandColorValues[c]!.fg, startsWith('#'));
      }
      expect(bandColorValues['gold']!.fg, '#1A2E1A');
      expect(bandColorValues['violet']!.fg, '#FFFFFF');
    });

    test('un color inventado cae a la neutra en vez de dejar la banda sin pintar', () {
      expect(bandOf('fucsia').bg, bandColorValues['neutral']!.bg);
      expect(bandOf(null).bg, bandColorValues['neutral']!.bg);
    });
  });

  group('el perfil', () {
    test('una persona llega con la misma forma y sin comunidad', () {
      final p = PublisherProfile.fromJson(const {
        'publisherId': 'u1',
        'slug': 'luis',
        'displayName': 'Luis',
        'kind': 'creator',
        'followersCount': 0,
        'isFollowing': false,
        'albumsCount': 2,
        'editionsCount': 5,
      });
      expect(p.esPersona, isTrue);
      expect(p.canManage, isFalse);
      // 0 seguidores aquí es un VALOR, no un hueco: no hay a quién seguir.
      expect(p.followersCount, 0);
      expect(p.isFollowing, isFalse);
    });

    test('álbumes y ediciones son cifras distintas', () {
      final p = PublisherProfile.fromJson(const {
        'publisherId': 'p1',
        'slug': 'fe',
        'displayName': 'Forjabrasa',
        'albumsCount': 2,
        'editionsCount': 7,
      });
      expect(p.albumsCount, 2);
      expect(p.editionsCount, 7);
    });

    test('sin lastPostAt NO es un error: es el perfil recién creado', () {
      final p = PublisherProfile.fromJson(const {
        'publisherId': 'p1',
        'slug': 'fe',
        'displayName': 'Forjabrasa',
      });
      expect(p.lastPostAt, isNull);
      expect(p.bandColor, isNull);
      expect(bandOf(p.bandColor).bg, bandColorValues['neutral']!.bg);
    });

    test('un perfil privado llega sin descripción y NO está roto', () {
      // El servidor degrada el nombre al username y esconde la bio, pero el
      // catálogo sigue llegando: se oculta la persona, no su obra.
      final p = PublisherProfile.fromJson(const {
        'publisherId': 'u2',
        'slug': 'mvidal',
        'displayName': 'mvidal',
        'kind': 'creator',
        'albumsCount': 3,
      });
      expect(p.description, isNull);
      expect(p.albumsCount, 3);
    });
  });

  group('validación', () {
    test('el nombre es obligatorio y tiene tope', () {
      expect(validatePublisherProfile(displayName: 'Forjabrasa').valid, isTrue);
      expect(validatePublisherProfile(displayName: '   ').valid, isFalse);
      expect(
        validatePublisherProfile(displayName: 'x' * (profileLimits.displayName.max + 1)).valid,
        isFalse,
      );
    });

    test('el lema es más corto que la descripción, y es la razón de que existan los dos', () {
      expect(profileLimits.tagline.max, 80);
      expect(profileLimits.description.max, 600);
      expect(
        validatePublisherProfile(
          displayName: 'F',
          tagline: 'x' * profileLimits.tagline.max,
        ).valid,
        isTrue,
      );
      final r = validatePublisherProfile(
        displayName: 'F',
        tagline: 'x' * (profileLimits.tagline.max + 1),
      );
      expect(r.valid, isFalse);
      expect(r.issues.single.field, 'tagline');
    });

    test('un color fuera del conjunto se rechaza', () {
      final r = validatePublisherProfile(displayName: 'F', bandColor: 'fucsia');
      expect(r.valid, isFalse);
      expect(r.issues.any((i) => i.field == 'bandColor'), isTrue);
    });
  });

  group('perfil a medio hacer', () {
    test('vacío es válido, pero se sabe que está vacío', () {
      expect(profileIsBare(), isTrue);
      expect(profileIsBare(tagline: '   '), isTrue);
      expect(profileIsBare(tagline: 'Cromos de bestiario'), isFalse);
      expect(profileIsBare(logoKey: 'x/y.webp'), isFalse);
    });
  });
}

File: packages/core/tests/publisher-profile.test.ts
/**
 * KRO-285 — contrato del perfil público de un publisher.
 *
 * Lo que se fija aquí es la división que gobierna el tipo: qué escribe el
 * publisher y qué pone el servidor. Y que un perfil VACÍO es válido — es el
 * estado de todo publisher recién creado, no un error.
 */
import { describe, it, expect } from 'vitest';
import {
  PROFILE_BAND_COLORS,
  BAND_COLOR_VALUES,
  DEFAULT_BAND_COLOR,
  PROFILE_LIMITS,
  isBandColor,
  validatePublisherProfile,
  profileIsBare,
} from '../src/publisher-profile';

const base = { displayName: 'Cinabrio' };

describe('KRO-285 · perfil público del publisher', () => {
  describe('el color de la banda', () => {
    it('es un conjunto CERRADO, no un hex libre', () => {
      expect(isBandColor('green')).toBe(true);
      expect(isBandColor('#ff00ff')).toBe(false);
      expect(isBandColor('fucsia')).toBe(false);
      expect(isBandColor(undefined)).toBe(false);
    });

    it('cada color tiene su valor Y su color de texto', () => {
      // El `fg` va en el contrato para garantizar que el nombre se LEA sobre la
      // banda. Dejarlo al criterio de cada host es cómo aparece blanco sobre oro.
      for (const c of PROFILE_BAND_COLORS) {
        const v = BAND_COLOR_VALUES[c];
        expect(v.bg).toMatch(/^#[0-9A-F]{6}$/i);
        expect(v.fg).toMatch(/^#[0-9A-F]{6}$/i);
        expect(v.bg).not.toBe(v.fg);
      }
    });

    it('el default es la banda de quien no ha elegido, y existe', () => {
      expect(isBandColor(DEFAULT_BAND_COLOR)).toBe(true);
    });
  });

  describe('validación', () => {
    it('con nombre basta: TODO lo demás es opcional', () => {
      // Un perfil recién creado no tiene nada, y es el estado normal de
      // cualquier publisher que acaba de darse de alta.
      expect(validatePublisherProfile(base).valid).toBe(true);
    });

    it('el nombre es obligatorio y acotado', () => {
      expect(validatePublisherProfile({}).valid).toBe(false);
      expect(validatePublisherProfile({ displayName: '   ' }).valid).toBe(false);
      expect(validatePublisherProfile({
        displayName: 'x'.repeat(PROFILE_LIMITS.displayName.max + 1),
      }).valid).toBe(false);
    });

    it('el lema es de UNA línea, y el tope es lo que lo hace serlo', () => {
      expect(PROFILE_LIMITS.tagline.max).toBeLessThan(PROFILE_LIMITS.description.max);
      const r = validatePublisherProfile({ ...base, tagline: 'x'.repeat(PROFILE_LIMITS.tagline.max + 1) });
      expect(r.valid).toBe(false);
      expect(r.issues[0].field).toBe('tagline');
    });

    it('acota descripción y ciudad', () => {
      expect(validatePublisherProfile({ ...base, description: 'x'.repeat(601) }).valid).toBe(false);
      expect(validatePublisherProfile({ ...base, city: 'x'.repeat(61) }).valid).toBe(false);
    });

    it('rechaza un color de banda inventado', () => {
      const r = validatePublisherProfile({ ...base, bandColor: 'fucsia' as any });
      expect(r.valid).toBe(false);
      expect(r.issues[0].field).toBe('bandColor');
    });

    it('IGNORA los calculados en vez de rechazarlos', () => {
      // Un cliente que reenvíe el objeto entero que acaba de leer es un cliente
      // razonable; hacerle fallar por eso solo le obliga a limpiar a mano.
      const conCalculados = {
        ...base, followersCount: 9999, albumsCount: 40, joinedAt: '2020-01-01T00:00:00Z',
      } as any;
      expect(validatePublisherProfile(conCalculados).valid).toBe(true);
    });
  });

  describe('profileIsBare — señal, no validación', () => {
    it('un perfil sin lema, descripción ni logo está pelado', () => {
      expect(profileIsBare(base)).toBe(true);
      expect(profileIsBare(null)).toBe(true);
      expect(profileIsBare({ ...base, tagline: '   ' })).toBe(true);
    });

    it('con cualquiera de los tres, ya no', () => {
      expect(profileIsBare({ ...base, tagline: 'Cromos de bestias' })).toBe(false);
      expect(profileIsBare({ ...base, description: 'Hola' })).toBe(false);
      expect(profileIsBare({ ...base, logoKey: 'publishers/cinabrio/logo.webp' })).toBe(false);
    });

    it('pelado NO significa inválido: son preguntas distintas', () => {
      expect(profileIsBare(base)).toBe(true);
      expect(validatePublisherProfile(base).valid).toBe(true);
    });
  });
});
